File: kata-solutions/countries-kata/countries-kata-python/src/country_list.py
from country import Country
from math import sqrt
from ports_adapters import CsvCountriesOutputAdapter, RestCountriesInputAdapter, CountriesDTO

class CountryList:
  def __init__(self, 
               input_port = RestCountriesInputAdapter(),
               output_port = CsvCountriesOutputAdapter()):
    self._countries = input_port.load_all()
    self._output_port = output_port
  
  def sorted_by_population(self):
    return sorted(self._countries, key=lambda x: getattr(x, 'population'))

  def average_population(self):
    return average_of([country.population for country in self._countries])

  def standard_deviation(self):
    return standard_deviation_of([country.population for country in self._countries])
  
  def standard_deviations_per_country(self):
    standard_deviation = self.standard_deviation()
    return [
      round(abs(self.average_population() - country.population) / standard_deviation, 2) 
      for country in self.sorted_by_population()]

  def export_to_csv(self):
    sorted_countries = self.sorted_by_population()
    standard_deviations_per_country = self.standard_deviations_per_country()
    countries_dto = CountriesDTO([[sorted_countries[i].name, 
            sorted_countries[i].capital, 
            sorted_countries[i].population, 
            standard_deviations_per_country[i]] for i in range(len(self._countries))])
    self._output_port.export(countries_dto)
    
def average_of(a_collection):
  if not a_collection: return 0
  return sum([item for item in a_collection]) / len(a_collection)

def standard_deviation_of(a_collection):
  if not a_collection: return 0
  return sqrt(sum([(item - average_of(a_collection)) ** 2 for item in a_collection]) / len(a_collection))

def main():
    CountryList().export_to_csv()
    print("List with country data exported to countries.csv")

if __name__ == "__main__":
    main()

File: kata-solutions/countries-kata/countries-kata-typescript/src/CountryList.ts
'use strict';

import { InputPort, OutputPort } from "./Ports";
import { Country} from "./Country";
import { RestCountriesInputAdapter } from "./RestCountriesInputAdapter";
import { CsvOutputAdapter } from "./CsvOutputAdapter";

export class CountryList {
  private countryList: Country[];
  private outputPort: OutputPort;
  
   private constructor(inputPort: InputPort, outputPort: OutputPort) {
    this.countryList = inputPort.load_all();
    this.outputPort = outputPort;
  }

  static async create_instance(inputPort: InputPort, outputPort: OutputPort = new CsvOutputAdapter()): Promise<CountryList> {
     return new CountryList(inputPort, outputPort);
  }

  public compare(country1: Country, country2: Country): number {
    return country1.population - country2.population
  }

  public sorted_by_population(): Country[] {
    return this.countryList.sort(this.compare);
  }

  public sumOf(anArray: number[]): number {
    return anArray.reduce((total, item) => total + item, 0);
  }

  public average_population() {
    if (!this.countryList.length)
      return 0;

    var totalPopulation = this.sumOf(this.countryList.map(country => country.population));
    return totalPopulation / this.countryList.length;
  }

  public standard_deviation(): number {
    if (!this.countryList.length)
      return 0;

    var average = this.average_population();
    var diviationsList = this.countryList.map(country => (country.population - average) * (country.population - average));
    return Math.sqrt(this.sumOf(diviationsList) / this.countryList.length)
  }

  public standard_deviations_for(country: Country): number {
    var standardDeviations = Math.abs(this.average_population() - country.population) / this.standard_deviation();
    return Math.round((standardDeviations + Number.EPSILON) * 100) / 100
  }

  public as_nested_array(): string[][] {
    var sortedCountries = this.sorted_by_population();
    return sortedCountries.map(country => [country.name, country.capital, String(country.population), String(this.standard_deviations_for(country))]);
  }

  to_csv() {
    this.outputPort.write(this.as_nested_array());
  }
}


async function main() {
  let countryList = await CountryList.create_instance(await RestCountriesInputAdapter.instance());
  countryList.to_csv();
}

if (require.main === module) {
  main();
}